src/canvasdrawer.js

```diff
@@ -0,0 +1,1151 @@
+ * met:
+ *
+ * - Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ *
+ * - Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in the
+ *   documentation and/or other materials provided with the distribution.
+ *
+ * - Neither the name of CodePlex Foundation nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+(function( $ ){
+
+/**
+ * @class CanvasDrawer
+ * @memberof OpenSeadragon
+ * @classdesc Default implementation of CanvasDrawer for an {@link OpenSeadragon.Viewer}.
+ * @param {Object} options - Options for this Drawer.
+ * @param {OpenSeadragon.Viewer} options.viewer - The Viewer that owns this Drawer.
+ * @param {OpenSeadragon.Viewport} options.viewport - Reference to Viewer viewport.
+ * @param {Element} options.element - Parent element.
+ * @param {Number} [options.debugGridColor] - See debugGridColor in {@link OpenSeadragon.Options} for details.
+ */
+
+class CanvasDrawer extends $.DrawerBase{
+    constructor(){
+        super(...arguments);
+
+        // Options
+
+
+
+        /**
+         * 2d drawing context for {@link OpenSeadragon.Drawer#canvas} if it's a &lt;canvas&gt; element, otherwise null.
+         * @member {Object} context
+         * @memberof OpenSeadragon.Drawer#
+         */
+        this.context = this.canvas.getContext( '2d' );
+
+        /**
+         * Sketch canvas used to temporarily draw tiles which cannot be drawn directly
+         * to the main canvas due to opacity. Lazily initialized.
+         */
+        this.sketchCanvas = null;
+        this.sketchContext = null;
+
+        // We force our container to ltr because our drawing math doesn't work in rtl.
+        // This issue only affects our canvas renderer, but we do it always for consistency.
+        // Note that this means overlays you want to be rtl need to be explicitly set to rtl.
+        this.container.dir = 'ltr';
```

Comment:
This should either only be in DrawerBase or only in the implementations that need it. If the latter, do you know which implementations that is?

Reply:
Well, the existing comment said it was only relevant for `canvas`. It also said it was always set for consistency. I don't know if this is something we need to maintain moving forward...

I don't really know how to test this, I'm not sure what artifacts to expect. When I set it to `rtl` on the demo in the documentation pages it doesn't seem to break anything as far as I can tell.

I can remove it from `CanvasDrawer` for now and leave it in `DrawerBase` only though.

Reply:
Yeah, I guess one would need to set up an `rtl` page to test in. We could also track down the original commit to see if there are any clues there. If we can prove that it's just needed in the canvas renderer (and confirm that it's still needed at all), it might be good to limit it to the canvas renderer so we can start to phase it out. On the other hand, without knowing more about it, it might be safest to leave it in `DrawerBase`. I don't know... It's kind of tangential to the thrust of this patch, so I don't want to get too distracted by it.

Reply:
Okay, I tracked it down... It was me! 

https://github.com/openseadragon/openseadragon/issues/186

There are instructions there for how to repro the issue. Can you remove the `ltr` fix and try those repro steps? I wouldn't be surprised if it was just an issue with the older browsers. If so, we can remove it entirely.

Reply:
It's still an issue. It affects both the `canvas` and `webgl` drawers. I'll leave it in `DrawerBase`.

Reply:
Good to know! Maybe mention in the comment that it also affects the WebGL renderer? Is there anything else that should be added to that comment so we're not confused about it next time?

Reply:
Digging further, at least currently, it has nothing to do with the math regarding tile positioning as far as I can tell (unclear whether it used to...). Rather, the `<canvas>` element was positioned wrong when `dir='rtl'`. Adding `canvas.left = 0;` fixes the issue for all drawer types.

Reply:
Nice! That seems like a cleaner solution. It would be good to comment it to make it clear why we're doing it.